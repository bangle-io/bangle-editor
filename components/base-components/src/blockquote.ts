import type { MarkdownSerializerState } from 'prosemirror-markdown';

import type { RawPlugins, RawSpecs } from '@bangle.dev/core';
import {
  Command,
  DOMOutputSpec,
  EditorState,
  keymap,
  Node,
  Schema,
  wrapIn,
  wrappingInputRule,
} from '@bangle.dev/pm';
import {
  copyEmptyCommand,
  cutEmptyCommand,
  moveNode,
} from '@bangle.dev/pm-commands';
import {
  createObject,
  filter,
  findParentNodeOfType,
  insertEmpty,
} from '@bangle.dev/utils';

import { getNodeType, getParaNodeType } from './helpers';

export const spec = specFactory;
export const plugins = pluginsFactory;
export const commands = {
  queryIsBlockquoteActive,
  wrapInBlockquote,
};
export const defaultKeys = {
  wrapIn: 'Ctrl-ArrowRight',
  moveDown: 'Alt-ArrowDown',
  moveUp: 'Alt-ArrowUp',
  emptyCopy: 'Mod-c',
  emptyCut: 'Mod-x',
  insertEmptyParaAbove: 'Mod-Shift-Enter',
  insertEmptyParaBelow: 'Mod-Enter',
};

const name = 'blockquote';

function specFactory(): RawSpecs {
  return {
    type: 'node',
    name,
    schema: {
      content: 'block*',
      group: 'block',
      defining: true,
      draggable: false,
      parseDOM: [{ tag: 'blockquote' }],
      toDOM: (): DOMOutputSpec => {
        return ['blockquote', 0];
      },
    },
    markdown: {
      toMarkdown: (state: MarkdownSerializerState, node: Node) => {
        state.wrapBlock('> ', null, node, () => state.renderContent(node));
      },
      parseMarkdown: {
        blockquote: {
          block: name,
        },
      },
    },
  };
}

function pluginsFactory({
  markdownShortcut = true,
  keybindings = defaultKeys,
} = {}): RawPlugins {
  return ({ schema }) => {
    const type = getNodeType(schema, name);
    return [
      markdownShortcut && wrappingInputRule(/^\s*>\s$/, type),
      keybindings &&
        keymap(
          createObject([
            [keybindings.wrapIn, wrapInBlockquote()],
            [keybindings.moveUp, moveNode(type, 'UP')],
            [keybindings.moveDown, moveNode(type, 'DOWN')],

            [keybindings.emptyCopy, copyEmptyCommand(type)],
            [keybindings.emptyCut, cutEmptyCommand(type)],

            [keybindings.insertEmptyParaAbove, insertEmptyParaAbove()],
            [keybindings.insertEmptyParaBelow, insertEmptyParaBelow()],
          ]),
        ),
    ];
  };
}

export function queryIsBlockquoteActive() {
  return (state: EditorState) => {
    const type = getNodeType(state, name);
    return Boolean(findParentNodeOfType(type)(state.selection));
  };
}

export function wrapInBlockquote() {
  return filter(
    (state) => !queryIsBlockquoteActive()(state),
    (state, dispatch, _view) => {
      const type = getNodeType(state, name);

      return wrapIn(type)(state, dispatch);
    },
  );
}

export function insertEmptyParaAbove(): Command {
  const isInBlockquote = queryIsBlockquoteActive();
  return (state, dispatch, view) => {
    const para = getParaNodeType(state);

    return filter(isInBlockquote, insertEmpty(para, 'above', true))(
      state,
      dispatch,
      view,
    );
  };
}

export function insertEmptyParaBelow(): Command {
  const isInBlockquote = queryIsBlockquoteActive();
  return (state, dispatch, view) => {
    const para = getParaNodeType(state);

    return filter(isInBlockquote, insertEmpty(para, 'below', true))(
      state,
      dispatch,
      view,
    );
  };
}
