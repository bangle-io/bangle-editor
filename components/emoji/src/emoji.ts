import type { MarkdownSerializerState } from 'prosemirror-markdown';

import { domSerializationHelpers, RawSpecs } from '@bangle.dev/core';
import type { Command, Node } from '@bangle.dev/pm';

export const spec = specFactory;
export const plugins = pluginsFactory;
export const commands = {
  insertEmoji,
};

const name = 'emoji';

const getTypeFromSchema = (schema: any) => schema.nodes[name];

function specFactory({
  getEmoji,
  defaultEmojiAlias = 'smiley',
}: {
  getEmoji: (alias: string, node: Node) => string;
  defaultEmojiAlias?: string;
}): RawSpecs {
  const { toDOM, parseDOM } = domSerializationHelpers(name, {
    tag: 'span',
    parsingPriority: 51,
    content: (node) => {
      let result = getEmoji(node.attrs['emojiAlias'], node);
      return result;
    },
  });

  return {
    type: 'node',
    name,
    schema: {
      attrs: {
        emojiAlias: {
          default: defaultEmojiAlias,
        },
      },
      inline: true,
      group: 'inline',
      draggable: true,
      atom: true,
      toDOM,
      parseDOM,
    },

    markdown: {
      toMarkdown: (state: MarkdownSerializerState, node: Node) => {
        state.write(`:${node.attrs['emojiAlias']}:`);
      },
      parseMarkdown: {
        emoji: {
          node: 'emoji',
          getAttrs: (tok: any) => {
            return {
              emojiAlias: tok.markup,
            };
          },
        },
      },
    },
  };
}

function pluginsFactory({ keybindings = {} } = {}) {
  return () => {
    return [];
  };
}

export function insertEmoji(emojiAlias: string): Command {
  return function (state, dispatch) {
    let emojiType = getTypeFromSchema(state.schema);

    let { $from } = state.selection,
      index = $from.index();
    if (!$from.parent.canReplaceWith(index, index, emojiType)) {
      return false;
    }
    if (dispatch) {
      const attr = {
        emojiAlias: emojiAlias,
      };

      dispatch(state.tr.replaceSelectionWith(emojiType.create(attr)));
    }
    return true;
  };
}
