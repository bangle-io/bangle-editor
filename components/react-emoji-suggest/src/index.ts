import * as emojiSuggest from './emoji-suggest';
export * from './EmojiSuggest';
export { emojiSuggest };

export type { GetEmojiGroupsType } from './emoji-suggest';
export { selectEmoji } from './emoji-suggest';
export { getSquareDimensions, resolveCounter, resolveRowJump } from './utils';
