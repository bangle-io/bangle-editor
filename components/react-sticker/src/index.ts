import * as sticker from './sticker';

export { sticker };
