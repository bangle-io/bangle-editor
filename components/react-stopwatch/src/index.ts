import * as stopwatch from './stopwatch';

export { stopwatch };
