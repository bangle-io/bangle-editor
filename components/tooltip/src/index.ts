import * as selectionTooltip from './selection-tooltip';
import type { SuggestTooltipRenderOpts } from './suggest-tooltip';
import * as suggestTooltip from './suggest-tooltip';
import type { TooltipRenderOpts } from './tooltip-placement';
import * as tooltipPlacement from './tooltip-placement';

export * from './create-tooltip-dom';
export { tooltipPlacement, selectionTooltip, suggestTooltip };
export type { TooltipRenderOpts };
export type { SuggestTooltipRenderOpts };
export type { SelectionTooltipProps } from './selection-tooltip';
