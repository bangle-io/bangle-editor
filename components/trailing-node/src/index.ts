import * as trailingNode from './trailing-node';

export { trailingNode };
