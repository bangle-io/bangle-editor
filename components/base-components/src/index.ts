// marks
import * as bold from './bold';
import * as code from './code';
import * as italic from './italic';
import * as link from './link';
import * as strike from './strike';
import * as underline from './underline';

// nodes
import * as blockquote from './blockquote';
import * as bulletList from './bullet-list';
import * as codeBlock from './code-block';
import * as hardBreak from './hard-break';
import * as heading from './heading';
import * as horizontalRule from './horizontal-rule';
import * as listItem from './list-item/list-item-component';
import * as orderedList from './ordered-list';
import * as image from './image';

export {
  doc,
  paragraph,
  text,
  history,
  editorStateCounter,
} from '@bangle.dev/core';

export {
  bold,
  code,
  italic,
  link,
  strike,
  underline,
  blockquote,
  bulletList,
  codeBlock,
  hardBreak,
  heading,
  horizontalRule,
  listItem,
  orderedList,
  image,
};
