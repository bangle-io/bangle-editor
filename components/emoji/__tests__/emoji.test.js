/**
 * @jest-environment jsdom
 */
/** @jsx psx */
import emojiParser from 'markdown-it-emoji';

import { defaultPlugins, defaultSpecs } from '@bangle.dev/all-base-components';
import { SpecRegistry } from '@bangle.dev/core';
import {
  getDefaultMarkdownItTokenizer,
  markdownParser,
  markdownSerializer,
} from '@bangle.dev/markdown';
import { psx, renderTestEditor } from '@bangle.dev/test-helpers';

import { emoji } from '../src/index';

let getEmoji, testEditor, specRegistry;

beforeEach(() => {
  getEmoji = jest.fn(() => 'horse');

  specRegistry = new SpecRegistry([
    ...defaultSpecs(),
    emoji.spec({ getEmoji }),
  ]);
  const plugins = [...defaultPlugins(), emoji.plugins()];

  testEditor = renderTestEditor({
    specRegistry,
    plugins,
  });
});

test('Rendering works', async () => {
  Date.now = jest.fn(() => 0);
  const { container, view } = testEditor(
    <doc>
      <para>
        foo[]bar
        <emoji emojiAlias="horse" />
      </para>
    </doc>,
  );

  expect(view.state).toEqualDocAndSelection(
    <doc>
      <para>
        foo[]bar
        <emoji emojiAlias="horse" />
      </para>
    </doc>,
  );
  expect(getEmoji).toMatchInlineSnapshot(`
    [MockFunction] {
      "calls": [
        [
          "horse",
          {
            "attrs": {
              "emojiAlias": "horse",
            },
            "type": "emoji",
          },
        ],
      ],
      "results": [
        {
          "type": "return",
          "value": "horse",
        },
      ],
    }
  `);
  expect(
    container.querySelector(`[data-bangle-name="emoji"]`),
  ).toMatchSnapshot();
});

describe('markdown', () => {
  const serialize = async (doc) => {
    let content = doc;
    if (typeof doc === 'function') {
      content = doc(specRegistry.schema);
    }
    return markdownSerializer(specRegistry).serialize(content);
  };
  const parse = async (md) =>
    markdownParser(
      specRegistry,
      getDefaultMarkdownItTokenizer().use(emojiParser),
    ).parse(md);

  test('markdown 1', async () => {
    const doc = (
      <doc>
        <para>
          hello world
          <emoji emojiAlias="horse" />
        </para>
      </doc>
    );
    const md = await serialize(doc);
    expect(md).toMatchInlineSnapshot(`"hello world:horse:"`);

    expect(await parse(md)).toEqualDocument(doc);
  });

  test('markdown 2', async () => {
    const doc = (
      <doc>
        <para>
          hello world
          <emoji />
        </para>
      </doc>
    );
    const md = await serialize(doc);
    expect(md).toMatchInlineSnapshot(`"hello world:smiley:"`);
    expect(await parse(md)).toEqualDocument(
      <doc>
        <para>
          hello world
          <emoji />
        </para>
      </doc>,
    );
  });
});
