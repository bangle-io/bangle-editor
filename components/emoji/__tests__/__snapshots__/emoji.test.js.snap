// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering works 2`] = `
<span
  data-bangle-attrs="{"emojiAlias":"horse"}"
  data-bangle-name="emoji"
  draggable="true"
>
  horse
</span>
`;
